x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -357,18 +368,122 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
+                    client.threadPool()
+                        .schedule(
+                            () -> getOrCreateProfileWithBackoff(subject, profileDocument, backoff, listener),
+                            backoffTimeValue,
+                            ThreadPool.Names.GENERIC
+                        );
+                } else {
+                    // Retry has depleted. This can only happen when the document or the profile index itself gets deleted
+                    // in between requests.
+                    listener.onFailure(
+                        new ElasticsearchException("failed to retrieving profile [{}] after all retries", profileDocument.uid())
+                    );
+                }
+                return;
+            }
+            if (isSubjectOwnerOfProfileDocument(subject, versionedDocument.doc)) {
+                // The profile document is a match. It must have just got created by another thread, i.e. racing.
+                // Still need to update it with current auth info before return
+                updateProfileForActivate(subject, versionedDocument, listener);
+            } else {
+                // The profile document is NOT a match, this means either genuine hash collision or profile document
+                // was manually updated, e.g. realm or user rename etc. So we attempt to differentiate from the
+                // existing profile document by increase the differentiator number by 1.
+                incrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
+            }
+        }, listener::onFailure));
+    }
+
+    private void incrementDifferentiatorAndCreateNewProfile(
+        Subject subject,
+        ProfileDocument profileDocument,
+        ActionListener<Profile> listener
+    ) throws IOException {
+        final String uid = profileDocument.uid();
+        final int index = uid.lastIndexOf("_");
+        if (index == -1) {
+            throw new ElasticsearchException("profile uid [{}] does not contain any underscore character", uid);
+        }
+        final String baseUid = uid.substring(0, index);
+        final String differentiatorString = uid.substring(index + 1);
+        if (differentiatorString.isBlank()) {
+            throw new ElasticsearchException("profile uid [{}] does not contain a differentiator", uid);
+        }
+        final int differentiator;
+        try {
+            differentiator = Integer.parseInt(differentiatorString);
+        } catch (NumberFormatException e) {
+            throw new ElasticsearchException("profile uid [{}] differentiator is not a number", e, uid);
+        }
+        // Prevent infinite recursion. It is practically impossible to get this many clashes
+        if (differentiator > DIFFERENTIATOR_UPPER_LIMIT) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator number is too high", profileDocument.uid()));
+            return;
+        }
```

Comment:
The upper limit of differentiator is 9 (0-based) which means it is not possible to have more than 10 profile documents sharing the same base UID. It is impractical to reach this number considering how clash can happen. Nevertheless, when it does happen, the API throw an error which is a non-ideal but defintive behaviour which in my opinion is acceptable for now.